AMP-48: School CRUD Page

ksdfg opened the pull request.

ksdfg committed:
redux/school: add Basic actions, reducer and saga for school operations

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
components/DataPage: create a separate component for viewing data in a DataGrid along with a create/edit modal

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
redux: add school reducer and saga

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/schools: Add a page to handle School read, create and update operations

Use the DataPage component to render the page, while using a new component SchoolModal to handle create and delete operations. Also add route to /app/schools in the dashboard and a link in the sidebar.

Get current select school from school reducer, not user.

Signed-off-by: ksdfg <ksdfg123@gmail.com>

akhilnarang commented:
I know its still a draft, but just thought I'd note down whatever I faced

1. Enter details to create a new school, hit enter
2. Loader pops up, school gets created, modal still stays, with the values I just entered

I guess we should either clear the text, or dismiss it after the loader
And error handling, which I guess is planned anyway.

Faced this in the toast sometimes -
![image](https://user-images.githubusercontent.com/10119037/106279026-ca480780-6261-11eb-9bc9-71c4d7ca8c88.png)

I also managed to break CORS along the way :)

ksdfg committed:
pages/schools/components/SchoolModal: Update success modal to show school name correctly.

Schools have a name, not a full_name. Issue caused by confusion between School and User models

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg commented:
> I know its still a draft, but just thought I'd note down whatever I faced
> 
>     1. Enter details to create a new school, hit enter
> 
>     2. Loader pops up, school gets created, modal still stays, with the values I just entered
> 
> 
> I guess we should either clear the text, or dismiss it after the loader
> And error handling, which I guess is planned anyway.
> 
> Faced this in the toast sometimes -
> ![image](https://user-images.githubusercontent.com/10119037/106279026-ca480780-6261-11eb-9bc9-71c4d7ca8c88.png)
> 
> I also managed to break CORS along the way :)

Fixed the toast, silly mistake on my end.

As for the modal closing after operation is done, I actually wanted to try and do that for a while, but I couldn't figure out how to. I don't plan on adding any specific validation as of now, you guys can hopefully do that when you have time.

ksdfg committed:
pages/schools: close popup modal when the task is completed successfully

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/users/components/UserModal: add required fields for create user

ksdfg committed:
pages/users: close popup modal when the task is completed successfully

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/schools/components/SchoolModal: Add validations

Signed-off-by: ksdfg <ksdfg123@gmail.com>

PranavBakre commented on src/pages/schools/components/SchoolModal.jsx:
```diff
@@ -0,0 +1,164 @@
+            if (submit.hasOwnProperty("name") && submit["name"]) {
+                this.props.CreateSchool(submit);
+            } else {
+                toast.error("Please add email, type and password 😓", {
```
What is this for?

ksdfg replied:
hmm, copy pasting is never a good idea

PranavBakre replied:
also why hasOwnProperty?

ksdfg replied:
that entire part is now redundant tbh, since I'm validating all inputs now anyway. This was just a final check before submitting to make sure there is some form of validation available

PranavBakre commented:
> > I know its still a draft, but just thought I'd note down whatever I faced
> > ```
> > 1. Enter details to create a new school, hit enter
> > 
> > 2. Loader pops up, school gets created, modal still stays, with the values I just entered
> > ```
> > 
> > 
> > I guess we should either clear the text, or dismiss it after the loader
> > And error handling, which I guess is planned anyway.
> > Faced this in the toast sometimes -
> > ![image](https://user-images.githubusercontent.com/10119037/106279026-ca480780-6261-11eb-9bc9-71c4d7ca8c88.png)
> > I also managed to break CORS along the way :)
> 
> Fixed the toast, silly mistake on my end.
> 
> As for the modal closing after operation is done, I actually wanted to try and do that for a while, but I couldn't figure out how to. I don't plan on adding any specific validation as of now, you guys can hopefully do that when you have time.




you can use onClose directly in componentDidUpdate na? right under where toast.success is called

PranavBakre commented:
@PranavBakre ! ignore this. markdown reeped this before I had to update

ksdfg commented:
> > > I know its still a draft, but just thought I'd note down whatever I faced
> > > ```
> > > 1. Enter details to create a new school, hit enter
> > > 
> > > 2. Loader pops up, school gets created, modal still stays, with the values I just entered
> > > ```
> > > 
> > > 
> > > I guess we should either clear the text, or dismiss it after the loader
> > > And error handling, which I guess is planned anyway.
> > > Faced this in the toast sometimes -
> > > ![image](https://user-images.githubusercontent.com/10119037/106279026-ca480780-6261-11eb-9bc9-71c4d7ca8c88.png)
> > > I also managed to break CORS along the way :)
> > 
> > 
> > Fixed the toast, silly mistake on my end.
> > As for the modal closing after operation is done, I actually wanted to try and do that for a while, but I couldn't figure out how to. I don't plan on adding any specific validation as of now, you guys can hopefully do that when you have time.
> 
> you can use onClose directly in componentDidUpdate na? right under where toast.success is called

my bad, I forgot to update here, the modal now closes on update/create.
was done during the rush before the demo, hence forgot to update comment.

PranavBakre commented:
@ksdfg but why did u use componentDidUpdate when you had an closeModal

ksdfg commented:
@PranavBakre so as to not have modal state var as open in index.jsx iirc

ksdfg committed:
Update data to be displayed in /app/users, and modify DataPage component

pages/users:
- Add School column
- Hide Picture and Email columns by default
- Set Picture height to match row height dynamically

components/DataPage:
- Add Toolbar to DataGrid
- Change heading to h3 and change align it to center vertically within grid
- Remove custom row height

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
Simplify modal autoclose and fix bug on create for user and school

- Use `this.props.onClose()` in both SchoolModal and UserModal to close the modal after the success toast is given, instead of closing it from their respective parent components
- Use a `formSubmitted` state var to check if the change in selected user was due to form submission or simple selection of a new user/school

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/dashboard: Handle errors for all modules

Since there are multiple modules, with many pages that are accessible via dashboard potentially handling multiple modules, there is a risk for the same error to trigger multiple toasts if the errors were handled separately in each page.
Since the error messages are set per module, if we check for all error messages and render the respective error toasts in the root component itself (which is the dashboard component in this case) we can avoid the multiple toast situation with ease.
This also allows us to not have the headache of not having to set toast display for error messages in each child component of dashboard. However, in the future, if we need to make sure that a component does not trigger a toast for a certain module's error message, that can be done via a map of booleans set as a state variable in the dashboard component itself, with the child component switching the value for the specific module (essentially switching error toasts on or off).

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/users: Show school name instead of school ID in List of Users

Instead of just passing `this.props.users` i.e. fetched list of users from backend to DataPage component for rendering rows in DataGrid, use it to initialise a state variable `users`. Once we have fetched list of all schools, replace the `school` parameter of each object in `users` (which is initially set to school ID) to the name of the respective school.

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
components/DataPage: set title to h4

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/users/components/Usermodal: add dropdown to select school for user create/update

List of all schools should be fetched when parent component of modal was loaded; however, in case of some errors, fetch all schools if list of schools is empty.

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
pages/homepage: show school on homepage

Since there is no way to change school using Update User Me, just show the data.

Also instead of using disabled fields for non admin users, set onChange of textFields as null to make sure user cannot make any changes or submit form

Signed-off-by: ksdfg <ksdfg123@gmail.com>

PranavBakre committed:
@ksdfg remember to check console for errors

FIX: error is boolean in TextField
FIX: provide key in this.props.school.map, REASON: https://reactjs.org/docs/lists-and-keys.html#keys
Signed-off-by: Pranav Bakre <psbakre@yahoo.com>

ksdfg committed:
pages/homepage: add helpertext to school textfield to make sure users know that this cannot be edited from the homepage

Signed-off-by: ksdfg <ksdfg123@gmail.com>

PranavBakre commented on src/pages/dashboard/index.jsx:
```diff
@@ -36,10 +37,13 @@ class Dashboard extends Component {
-                position: toast.POSITION.TOP_CENTER,
-            });
+        // check if any new error message needs to be displayed
+        for (let i = 0; i < this.props.errorMessage.length; i++) {
```
Could you try array.forEach loop here. Checking with length looks weird when you have functions that can iterate. Array.foreach can also get you index.

ksdfg replied:
I did try with forEach before, but there was some issue iirc
let me try again now that I have slept a little

ksdfg replied:
k nvm, found a huge bug with this

If we trigger the same error multiple times, it only shows up the first time

ksdfg replied:
because technically the error did not change

PranavBakre replied:
wouldnt this happen right now as well?

PranavBakre replied:
Also why would multiple same errors not work? If I am right, every action that could leads to an error first sets errorMessage to ""

ksdfg replied:
I mean, I just tried, didn't get any errors

and yeah, that is happening right now as well

ksdfg replied:
now I'm not sure if it's because of my code or the damn cors issue

ksdfg replied:
hmmm, it seems to be working for create school and create user errors, so maybe this can be chalked up to no errors being set by the CORS errors...

PranavBakre commented on src/redux/reducer.js:
```diff
@@ -14,12 +15,13 @@ const persistConfig = {
     if (action.type === AuthActionTypes.LOGOUT) {
         storage.removeItem("persist:auth");
-        state.auth = state.user = undefined;
+        state.auth = state.user = state.school = undefined;
```
A thought. What if you retrieve _persist and clear state and set it back? This wouldnt require changes over time when new modules are added

https://github.com/VPPR/frontend/blob/9d9a2463a6f2b72a7be0aa62f995f7a0467f2077/src/redux/reducer.js#L25

ksdfg replied:
that's probably a good idea. Might make a `misc` PR for small changes like this one....

ksdfg committed:
pages/dashboard: use forEach to iterate through error messages

Signed-off-by: ksdfg <ksdfg123@gmail.com>

ksdfg committed:
components/SideBar: add schools link to navigation list

Signed-off-by: ksdfg <ksdfg123@gmail.com>

akhilnarang reviewed (approved):
Tested basic functionality, seems good to me!
